thread.c

```diff
@@ -1903,6 +1920,16 @@ rb_threadptr_pending_interrupt_check_mask(rb_thread_t *th, VALUE err)
     for (i=0; i<mask_stack_len; i++) {
         mask = mask_stack[mask_stack_len-(i+1)];
 
+        if (SYMBOL_P(mask)) {
+            /* do not match RUBY_FATAL_THREAD_KILLED etc */
+            if (err != rb_cInteger) {
```

Comment:
It seems surprising (and undocumented) that an "exception" key of `Integer` [and by extension, `Object`/`Kernel`/similar] to `handle_interrupt` affects behaviour -- https://github.com/ruby/ruby/actions/runs/5973661284/job/16206337391?pr=8278#step:9:128

This conditional maintains the existing behaviour, though I suspect the logic could be simplified in future to specifically treat the special exceptions as matching Exception, rather than Integer.